plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/FileLocationsImpl.xtend

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright (c) 2013 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtend.core.macro
+
+import com.google.inject.Inject
+import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.xtend.lib.annotations.Accessors
+import org.eclipse.xtend.lib.macro.file.FileLocations
+import org.eclipse.xtend.lib.macro.file.Path
+import org.eclipse.xtext.generator.IContextualOutputConfigurationProvider
+import org.eclipse.xtext.workspace.IProjectConfig
+import org.eclipse.xtext.workspace.IWorkspaceConfigProvider
+
+/**
+ * @author Sven Efftinge - Initial contribution and API
+ */
+class FileLocationsImpl implements FileLocations {
+	
+	@Inject @Accessors IWorkspaceConfigProvider projectInformationProvider
+	@Inject @Accessors IContextualOutputConfigurationProvider outputConfigurationProvider
+	@Accessors Resource context
+	
+	protected def IProjectConfig getProjectConfig(Path path) {
+		val string = path.getSegments().get(0)
+		val projectConfig = projectInformationProvider.getWorkspaceConfig(context).findProjectByName(string)
+		if (projectConfig == null) {
+			throw new IllegalArgumentException("The project '"+string+"' has not been configured.")
+		}
+		return projectConfig
+	}
+	
+	override Path getSourceFolder(Path path) {
+		getProjectSourceFolders(path).findFirst[sourceFolder| path.startsWith(sourceFolder)]
+	}
+
+	override Path getTargetFolder(Path path) {
+		val projectFolder = getProjectFolder(path)
+		if (projectFolder == null) {
+		 return null
+		}
+		val outputConfiguration = outputConfigurationProvider.getOutputConfigurations(context).head
+		val sourceFolder = getSourceFolder(path)
+		val outputFolder = if (sourceFolder == null) {
+			outputConfiguration.outputDirectory
+		} else {
+			outputConfiguration.getOutputDirectory(sourceFolder.segments.tail.join('/'))
```

Comment:
uri.deresolve keeps the last segment of the base uri as the first segment of the relative uri. Surprised me as well.

```
"/users/oehme/src".deresolve("/users/oehme") == "oehme/src
```